### PR TITLE
NeXus: use median oscillation width

### PR DESCRIPTION
**Pull request by benjaminhwilliams**
If the difference between the axis positions of the first and second images in a scan is not representative of the average oscillation width (perhaps because the goniometer is a little slow to reach a steady speed), then this first oscillation width will not be representative of the scan as a whole.  Instead, find the average oscillation width and use that to construct the `dxtbx.model.Scan` instead.

### CODE REVIEW

**Comment by graeme-winter** [comment]
> What we probably should really be using here is [AXISNAME_increment_set](https://manual.nexusformat.org/classes/base_classes/NXtransformations.html#nxtransformations-axisname-increment-set-field)
> 
> ```
> $ h5ls -r /dls/i19-2/data/2022/cm31144-2/test_crystal/test_01.nxs/entry/sample/transformations
> /kappa                   Dataset {1}
> /omega                   Dataset {1}
> /phi                     Dataset {1750}
> /phi_end                 Dataset {1750}
> /phi_increment_set       Dataset {1750}
> /sam_x                   Dataset {1}
> /sam_y                   Dataset {1}
> /sam_z                   Dataset {1}
> ```
> 
> I started adding support for this in #514 but obviously [not all the way there](https://github.com/cctbx/dxtbx/pull/514/files#diff-7893ccedbd87b17775442971310cb56202edad9b3e86dd0f7b19beff46eb3a18R113). I can't quite remember why it wasn't actually used - perhaps because we actually record an array of values rather than a single value (which I think the specification implies).

Using the readback value is perfectly reasonable - the fix described above sounds sane to use the mean oscillation width in preference to the width of the first image. In some cases the demand values may not be that representative of the actual rotation (which is ultimately what we care about)

**Comment by dagewa** [comment]
Is this a mean of a sample where most values are the same but there could be a few anomalous values in the sample? If so, maybe the median is better?

**Comment by graeme-winter** [comment]
> Is this a mean of a sample where most values are the same but there could be a few anomalous values in the sample? If so, maybe the median is better?

I had also wondered about mean of IQR, for same reason, but possibly maybe we are starting to argue about angel configuration.

**Comment by rjgildea** [comment]
> Is this a mean of a sample where most values are the same but there could be a few anomalous values in the sample? If so, maybe the median is better?

I was also going to suggest whether the median might be more appropriate here - we're assuming by definition that the goniometer speed is constant for the vast majority of the scan, so median is possibly more approriate. I guess using the median would probably also address one of the [failing tests](https://dev.azure.com/azure-dxtbx/dxtbx/_build/results?buildId=3476&view=logs&j=bbfc4e03-b3dc-52f1-d3f9-562e8cca4b84&t=fc34eecf-32f3-5a61-c195-251d92645545&l=1550) associated with this PR.

Would it be worth (e.g.) printing out a warning if we detect any significant non-uniformity in the osciallation values?

File: tests/nexus/test_build_dxtbx_models.py [hunk 1]
```diff
@@ -177,8 +177,8 @@ def test_get_dxtbx_scan(nxmx_example):
     scan = dxtbx.nexus.get_dxtbx_scan(sample, instrument.detectors[0])
     assert scan.get_num_images() == 10
     assert scan.get_image_range() == (1, 10)
-    assert scan.get_oscillation() == (0.0, 0.1)
-    assert scan.get_oscillation_range() == (0.0, 1.0)
+    np.testing.assert_allclose(scan.get_oscillation(), (0.0, 0.1))
+    np.testing.assert_allclose(scan.get_oscillation_range(), (0.0, 1.0))
```
Comment (rjgildea) [inline tests/nexus/test_build_dxtbx_models.py:181]: Is there a particular reason you didn't use `pytest.approx` here?
```suggestion
    assert pytest.approx(scan.get_oscillation()) == (0.0, 0.1)
    assert pytest.approx(scan.get_oscillation_range()) == (0.0, 1.0)
```

Comment (benjaminhwilliams) [reply tests/nexus/test_build_dxtbx_models.py:181]: None, I just forgot it exists!

File: src/dxtbx/nexus/__init__.py [hunk 2]
```diff
@@ -107,16 +107,14 @@ def get_dxtbx_scan(
 
     oscillation = (0, 0)
     if is_rotation:
+        start = scan_axis[0].to("degree").magnitude
         if scan_axis.end:
-            end = scan_axis.end[0]
+            step = np.median(scan_axis.end[()] - scan_axis[()]).to("degree").magnitude
         elif num_images > 1:
-            end = scan_axis[1]
+            step = np.median(np.diff(scan_axis)).to("degree").magnitude
```
Comment (rjgildea) [inline src/dxtbx/nexus/__init__.py:114]: `np.diff` appears to downcast to an `ndarray`, as demonstrated by the following warning and traceback:

```
$ dials.show $(dials.data get -q vmxi_thaumatin)/image_15799_master.h5
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
/dls/science/users/hko55533/software/cctbx/conda_base/lib/python3.9/site-packages/numpy/core/_asarray.py:171: UnitStrippedWarning: The unit of the quantity is stripped when downcasting to ndarray.
  return array(a, dtype, copy=False, order=order, subok=True)
Traceback (most recent call last):
  File "/dls/science/users/hko55533/software/cctbx/build/../modules/dials/src/dials/command_line/show.py", line 695, in <module>
    run()
  File "/dls/science/users/hko55533/software/cctbx/conda_base/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/dls/science/users/hko55533/software/cctbx/build/../modules/dials/src/dials/command_line/show.py", line 208, in run
    params, options = parser.parse_args(args=args, show_diff_phil=True)
  File "/dls/science/users/hko55533/software/cctbx/modules/dials/src/dials/util/options.py", line 867, in parse_args
    params, args = self._phil_parser.parse_args(
  File "/dls/science/users/hko55533/software/cctbx/modules/dials/src/dials/util/options.py", line 550, in parse_args
    importer = Importer(
  File "/dls/science/users/hko55533/software/cctbx/modules/dials/src/dials/util/options.py", line 202, in __init__
    self.unhandled = self.try_read_experiments_from_images(
  File "/dls/science/users/hko55533/software/cctbx/modules/dials/src/dials/util/options.py", line 279, in try_read_experiments_from_images
    experiments = ExperimentListFactory.from_filenames(
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/model/experiment_list.py", line 551, in from_filenames
    imageset = format_class.get_imageset(
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/FormatMultiImage.py", line 161, in get_imageset
    format_instance = cls.get_instance(filenames[0], **format_kwargs)
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/Format.py", line 274, in get_instance
    Class._current_instance_ = Class(filename, **kwargs)
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/FormatNXmxDLS.py", line 65, in __init__
    super().__init__(image_file, **kwargs)
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/FormatNXmx.py", line 30, in __init__
    super().__init__(image_file, **kwargs)
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/FormatHDF5.py", line 16, in __init__
    Format.__init__(self, image_file, **kwargs)
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/Format.py", line 168, in __init__
    self.setup()
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/Format.py", line 178, in setup
    self._start()
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/FormatNXmxDLS.py", line 68, in _start
    super()._start()
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/format/FormatNXmx.py", line 46, in _start
    self._scan_model = dxtbx.nexus.get_dxtbx_scan(nxsample, nxdetector)
  File "/dls/science/users/hko55533/software/cctbx/modules/dxtbx/src/dxtbx/nexus/__init__.py", line 114, in get_dxtbx_scan
    step = np.diff(scan_axis).median().to("degree").magnitude
AttributeError: Please report this error to dials-support@lists.sourceforge.net: 'numpy.ndarray' object has no attribute 'median'
```
Instead, "manually" calculating the differences appears to retain the units:
```suggestion
            step = np.median(scan_axis[1:] - scan_axis[:-1]).to("degree").magnitude
```

Comment (benjaminhwilliams) [reply src/dxtbx/nexus/__init__.py]: It's a consequence of [`h5py.Dataset` having an `__array__` method](https://github.com/h5py/h5py/blob/c98ba1165f9eb880dd588a2c8a6be180e6bae691/h5py/_hl/dataset.py#L1034-L1047), which we don't overload in the definition of `NXtransformationsAxis`.

**Comment by benjaminhwilliams** [comment]
In conversation with Christine Beavers (principal beamline scientist of I15 at Diamond Light Source, formerly of beamline 12.2.2 at the Advanced Light Source), it seems a deviation in oscillation widths of ≥ 10% from the median is a sensible threshold for a warning.  Christine describes an investigation of the stability of the 12.2.2 diffractometer scan axis:
> What we found is that the jitter on the steps was most dramatic on the first couple steps, as you expect, so we had a longer run up and down to achieve a stable velocity.  We essentially threw away a couple images at the beginning of the scan.  We never saw much more than 2-5% error on the step width after the lengthened run-up, so 10% should be pretty safe.

**Review by rjgildea — APPROVED** [review]
This looks good now, it seems sensible to use the median value here, and thanks adding the warning 👍

**Comment by benjaminhwilliams** [comment]
@ajames0064, for info, this will be in the next DIALS minor release, 3.11.  If you need it urgently, it may be possible to backport it as a point release on DIALS 3.10.

**Pull request merged**